openwisp_network_topology/api/serializers.py

```diff
@@ -22,15 +31,193 @@ def to_representation(self, data):
+        nodes.append(node.json(dict=True, original=False))
+    netjson = OrderedDict(
+        (
+            ('type', 'NetworkGraph'),
+            ('protocol', obj.protocol),
+            ('version', obj.version),
+            ('metric', obj.metric),
+            ('label', obj.label),
+            ('id', str(obj.id)),
+            ('parser', obj.parser),
+            ('organization', str(obj.organization.id)),
+            ('strategy', obj.strategy),
+            ('url', obj.url),
+            ('key', obj.key),
+            ('expiration_time', obj.expiration_time),
+            ('receive_url', get_receive_url(obj.pk, obj.key)),
+            ('published', obj.published),
+            ('created', obj.created),
+            ('modified', obj.modified),
+            ('nodes', nodes),
+            ('links', links),
+        )
+    )
+    return netjson
+
 
+class NetworkGraphRepresentation(object):
     def to_representation(self, obj):
-        return obj.json(dict=True)
+        """
+        Overriding the default represenation
+        of Topology object.
+        """
+        topo_data = get_representation_data(obj)
+        topo_data['receive_url'] = self.context['request'].build_absolute_uri(
+            topo_data['receive_url']
+        )
+        return topo_data
+
 
+class TopologySerializer(NetworkGraphRepresentation, ValidatedModelSerializer):
     class Meta:
-        model = swapper.load_model('topology', 'Topology')
+        model = Topology
         fields = '__all__'
+
+
+class BaseSerializer(FilterSerializerByOrgManaged, ValidatedModelSerializer):
+    pass
+
+
+class NetworkGraphSerializer(BaseSerializer):
+    """
+    NetJSON NetworkGraph.
+    """
+
+    def to_representation(self, obj):
+        if self.context['request'].method == 'POST':
+            serializer = TopologySerializer(
+                self.instance, context={'request': self.context['request']}
+            )
+            return serializer.data
+        return obj.json(dict=True)
+
+    class Meta:
+        model = Topology
+        fields = (
+            'label',
+            'organization',
+            'parser',
+            'strategy',
+            'key',
+            'expiration_time',
+            'url',
+            'published',
+        )
         list_serializer_class = NetworkCollectionSerializer
+        extra_kwargs = {'published': {'initial': True}}
+
+
+class NetworkGraphUpdateSerializer(NetworkGraphRepresentation, BaseSerializer):
+    class Meta:
+        model = Topology
+        fields = (
+            'label',
+            'organization',
+            'parser',
+            'strategy',
+            'key',
+            'expiration_time',
+            'url',
+            'published',
+        )
+
+    def validate_strategy(self, value):
+        if value == 'receive' and not self.initial_data.get('key'):
+            raise serializers.ValidationError(
+                _('a key must be specified when using RECEIVE strategy')
```

Comment:
So are we writing the first word of the messages in title case or not?
I see title case below but not here!

Reply:
ok, did it.